sdk/core/core/src/az_curl_adapter.c

```diff
@@ -10,27 +10,218 @@
 
 typedef struct {
   struct curl_slist * p_list;
-} az_headers_data;
+} az_curl_headers_list;
 
-AZ_CALLBACK_FUNC(az_headers_to_curl, az_headers_data *, az_pair_visitor)
+/**
+ * @brief writes a header key and value to a buffer as a 0-terminated string and using a separator
+ * span in between. Returns error as soon as any of the write operations fails
+ *
+ * @param writable_buffer
+ * @param p_header
+ * @param separator
+ * @return az_result
+ */
+az_result az_write_to_buffer(
+    az_span const writable_buffer,
+    az_pair * const p_header,
+    az_const_span separator) {
+  az_span_builder writer = az_span_builder_create(writable_buffer);
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, p_header->key));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, separator));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, p_header->value));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, AZ_STR("\0")));
+  return AZ_OK;
+}
+
+/**
+ * @brief allocate a buffer for a header. Then reads the az_pair header and writes a buffer. Then
+ * uses that buffer to set curl header. Header is set only if write operations were OK. Buffer is
+ * free after setting curl header.
+ *
+ * @param p_header
+ * @param p_headers
+ * @param separator
+ * @return az_result
+ */
+az_result az_add_header_to_curl_list(
+    az_pair * const p_header,
+    az_curl_headers_list * const p_headers,
+    az_const_span separator) {
+  // allocate a buffet for header
+  int16_t buffer_size = p_header->key.size + separator.size + p_header->value.size + 1;
+  uint8_t * const p_writable_buffer = (uint8_t * const)malloc(buffer_size);
+  if (p_writable_buffer == NULL) {
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+  char * buffer = (char *)p_writable_buffer;
+
+  // write buffer
+  az_span const writable_buffer
+      = (az_span const){ .begin = p_writable_buffer, .size = buffer_size };
+  az_result write_result = az_write_to_buffer(writable_buffer, p_header, separator);
+
+  // attach header only when write was OK
+  if (az_succeeded(write_result)) {
+    p_headers->p_list = curl_slist_append(p_headers->p_list, buffer);
```

Comment:
curl_slist_append may return `NULL` if it fails. We must handle this case.

Reply:
created new issue and label for this things.
I think those little details are super important to ship a high quality product as the one we want.
but also think this can be hold for later so we can speed up building a first preview-version.

issue created
https://github.com/Azure/azure-sdk-for-c/issues/117

Reply:
IMHO, this should fixed ASAP.